File: Python/trees_graphs/binary_search_tree.py
# Implementation of binary search tree with node structure.
from collections import deque


class Node(object):
    """

    Node structure for binary search tree.

    """
    def __init__(self, data, left=None, right=None, parent=None):
        self.data = data
        self.left = left
        self.right = right
        self.parent = parent

    def __str__(self):
        return "Node with data: %d" % self.data


class BinarySearchTree(object):
    """

    BinarySearchTree class implements a binary search tree.

    """
    def __init__(self, data):
        self.root = Node(data)

# PUBLIC METHODS
    def get_root(self):
        return self.root

    def successor(self, data):
        """

        Get the next largest data after given *data* by in-order traversal.

        """
        if data == self.maximum(self.root):
            return None
        node = self.lookup(data)
        if node.right is not None:
            return self.minimum(node.right)
        parent = node.parent
        while parent is not None and node is parent.right:
            node = parent
            parent = parent.parent
        return parent.data

    def predecessor(self, data):
        if data == self.minimum(self.root):
            return None
        node = self.lookup(data)
        if node.left is not None:
            return self.maximum(node.left)
        parent = node.parent
        while parent is not None and node is parent.left:
            node = parent
            parent = parent.parent
        return parent.data

    def minimum(self, node):
        """

        Retrieve the minimum data in the tree.

        """
        curr = node
        while curr.left is not None:
            curr = curr.left
        return curr.data

    def minimum_recursive(self, node):
        if node.left is not None:
            return self.minimum_recursive(node.left)
        else:
            return node.data

    def maximum(self, node):
        """

        Retrieve the maximum data in the tree.

        """
        curr = node
        while curr.right is not None:
            curr = curr.right
        return curr.data

    def maximum_recursive(self, node):
        if node.right is not None:
            return self.maximum_recursive(node.right)
        else:
            return node.data

    def lookup(self, data):
        return self._lookup(self.root, data)

    def insert(self, data):
        """

        Wrapper to insert operation.

        """
        if self.root is None:
            self.root = Node(data)
        else:
            new_node = Node(data)
            self._insert(self.root, new_node)
        return self.root

    def delete(self, data):
        """

        Remove node with data in the tree and return the root. if data node
        does not exist, return None.

        """
        # check root is not none
        if self.root is None:
            return None
        # lookup node to make sure it exists
        node = self.lookup(data)
        if node is None:
            return None
        # initialize shortcut vars
        parent = node.parent
        left = node.left
        right = node.right
        # case 1: 0 child node
        if left is None and right is None:
            # node is left child
            if parent.left == node:
                parent.left = None
            else:  # node is right child
                parent.right = None
            node = None
        # case 2: 1 child node
        elif right is None:  # node has a left child
            # node is a left child of its parent
            if parent.left is node:
                parent.left = left
                left.parent = parent
            else:  # node is a right child of its parent
                parent.right = left
                left.parent = parent
            node = None
        elif left is None:  # node has a right child
            # node is a right child of its parent
            if parent.right is node:
                parent.right = right
                right.parent = parent
            else:  # node is a left child of its parent
                parent.left = right
                right.parent = parent
            node = None
        # case 3: 2 child nodes
        else:
            temp_data = self._min_node(node).data
            self.delete(temp_data)
            node.data = temp_data
        # return root
        return self.root

    def print_data(self, node):
        print node.data,
        return

    def preorder(self, node, callback):
        if node is None:
            return
        callback(node)
        self.preorder(node.left, callback)
        self.preorder(node.right, callback)
        return

    def inorder(self, node, callback):
        if node is None:
            return
        self.inorder(node.left, callback)
        callback(node)
        self.inorder(node.right, callback)

    def inorder_iterative(self):
        stack = []
        curr = self.root
        while curr is not None or len(stack) > 0:
            if curr is not None:
                stack.append(curr)
                curr = curr.left
            else:
                curr = stack.pop()
                print curr.data,
                curr = curr.right
        return

    def postorder(self, node, callback):
        if node is None:
            return
        self.postorder(node.left, callback)
        self.postorder(node.right, callback)
        callback(node)

    def levelorder(self, node, callback):
        if node is None:
            return
        queue = deque([node])
        while len(queue) > 0:
            curr = queue.popleft()
            callback(curr)
            if curr.left is not None:
                queue.append(curr.left)
            if curr.right is not None:
                queue.append(curr.right)

# PRIVATE METHODS
    def _min_node(self, node):
        """

        Find the minimum data starting from node.

        """
        while node.left is not None:
            node = node.left
        return node

    def _lookup(self, node, data):
        if data == node.data:
            return node
        elif data < node.data:
            return self._lookup(node.left, data)
        else:
            return self._lookup(node.right, data)

    def _insert(self, root, node):
        """

        Insert node to the tree/sub-tree with root as *root*.

        """
        if node.data < root.data:
            # node goes left
            if root.left is None:
                root.left = node
                node.parent = root
            else:
                self._insert(root.left, node)
        else:
            # node goes right
            if root.right is None:
                root.right = node
                node.parent = root
            else:
                self._insert(root.right, node)
        return


if __name__ == "__main__":
    tree = BinarySearchTree(25)
    root = tree.get_root()
    tree.insert(15)
    tree.insert(50)
    tree.insert(10)
    tree.insert(22)
    tree.insert(35)
    tree.insert(70)
    tree.insert(4)
    tree.insert(12)
    tree.insert(18)
    tree.insert(24)
    tree.insert(31)
    tree.insert(44)
    tree.insert(66)
    tree.insert(90)
    tree.preorder(root, tree.print_data)
    print ""
    tree.inorder(root, tree.print_data)
    print ""
    tree.inorder_iterative()
    print ""
    tree.postorder(root, tree.print_data)
    print ""
    tree.levelorder(root, tree.print_data)
    print ""
    print tree.lookup(50)
    print tree.lookup(12)
    print tree.lookup(18)
    print tree.lookup(90)
    tree.preorder(root, tree.print_data)
    print ""
    print tree.successor(90)
    print tree.predecessor(90)
    print tree.minimum_recursive(tree.root)
    print tree.maximum_recursive(tree.root)

File: PHP/data_structures/binary_trees/binary_search_tree.php
<?php

/**
 * Implementation of a binary search tree.
 */
class Node {
    public $data;
    public $left;
    public $right;

    public function __construct($data, $left=NULL, $right=NULL) {
        $this->data = $data;
        $this->left = $left;
        $this->right = $right;
    }
}

class BinarySearchTree {
    private $root;

    public function __construct($data) {
        $this->root = new Node($data);
    }

    public function insert($data) {
        $newNode = new Node($data);
        if (empty($this->root)) {
            $this->root = $newNode;
            return $this->root;
        }
        $this->insertNode($this->root, $newNode);
        return $this->root;
    }

    protected function insertNode($parentNode, $newNode) {
        if ($newNode->data <= $parentNode->data) {
            if (empty($parentNode->left)) {
                $parentNode->left = $newNode;
            } else {
                $this->insertNode($parentNode->left, $newNode);
            }
        } else {
            if (empty($parentNode->right)) {
                $parentNode->right = $newNode;
            } else {
                $this->insertNode($parentNode->right, $newNode);
            }
        }
    }

    public function delete($data) {
        if (empty($this->root)) {
            return false;
        }
        $node = $this->findNode($data);
        // case 1: 0 children
        if (empty($node->left) && empty($node->right)) {
            return $this->deleteLeafNode($node, $this->root);
        } else if (empty($node->left) || empty($node->right)) { // case 2: 1 children
            return $this->deleteSingleChildNode($node, $this->root);
        } else { // case 3: 2 children
            // get right child's smallest node 
            $deleteNode = $this->findSmallestNode($node->right);
            // swap right child smallest node with $node.
            $this->swap($node, $deleteNode);
            echo "after swap: {$node->data} : {$deleteNode->data}" . PHP_EOL;
            // delete the $deleteNode which has 0 or 1 child only.
            if ($this->isLeafNode($deleteNode)) {
                if ($node->right == $deleteNode) {
                    $node->right = NULL;
                } else {
                    return $this->deleteLeafNode($deleteNode, $node->right);
                }
            } else {
                return $this->deleteSingleChildNode($deleteNode, $node->right);
            }
        }
    }

    private function swap(&$node1, &$node2) {
        $temp = $node1->data;
        $node1->data = $node2->data;
        $node2->data = $temp;
        return;
    }

    private function isLeafNode(&$node) {
        if (empty($node->left) && empty($node->right)) {
            return true;
        }
        return false;
    }

    private function deleteSingleChildNode(&$node, &$ancestorNode) {
        if (empty($node) ||
            (empty($node->left) && empty($node->right)) ||
            (!empty($node->left && !empty($node->right)))) {
            return false;
        }
        $parent = $this->_findParentNode($ancestorNode, $node);
        if ($node == $this->root) {
            if (!empty($node->left)) {
                $this->root = $node->left;
            } else{
                $this->root = $node->right;
            }
        } else if ($parent->left == $node) {
            if (!empty($node->left)) {
                $parent->left = $node->left;
            } else {
                $parent->left = $node->right;
            }
        } else if ($parent->right == $node) {
            if (!empty($node->left)) {
                $parent->right = $node->left;
            } else {
                $parent->right = $node->right;
            }
        }
        return true;
    }

    private function deleteLeafNode(&$node, &$ancestorNode) {
        if (!empty($node->left) || !empty($node->right) || empty($node)) {
            return false;
        }

        $parent = $this->_findParentNode($ancestorNode, $node);
        if ($node == $this->root) {
            $this->root = NULL;
        } else if($parent->left == $node) {
            $parent->left = NULL;
        } else {
            $parent->right = NULL;
        }
        return true;
    }    

    private function findSmallestNode(&$node) {
        if (empty($node)) {
            return NULL;
        }
        $curr = $node;
        while (!empty($curr->left)) {
            $curr = $curr->left;
        }
        return $curr;
    }

    private function _findParentNode($startNode, $node) {
        if (empty($startNode) || empty($node)) {
            return NULL;
        } else if ($startNode == $node) {

        }
        $prev = NULL;
        $curr = $startNode;

        while ($curr != NULL) {
            if ($curr->data == $node->data) {
                return $prev;
            }
            $prev = $curr;
            if ($node->data < $curr->data) {
                $curr = $curr->left;
            } else {
                $curr = $curr->right;
            }
        }
    }

    public function find($data) {
        if (empty($this->root)) {
            return false;
        }
        $found = false;
        $curr = $this->root;
        while ($curr != NULL) {
            if ($curr->data == $data) {
                $found = true;
                break;
            } else {
                if ($data < $curr->data) {
                    if ($curr->left != NULL) {
                        $curr = $curr->left;
                    } else {
                        $found = false;
                        break;
                    }
                } else {
                    if ($curr->right != NULL) {
                        $curr = $curr->right;
                    } else {
                        $found = false;
                        break;
                    }
                }
            }
        }
        return $found;
    }

    public function findNode($data) {
        return $this->_findNode($this->root, $data);
    }

    private function _findNode($node, $data) {
        if (empty($node)) {
            return NULL;
        }
        if ($node->data == $data) {
            return $node;
        } else if ($data < $node->data) {
            return $this->_findNode($node->left, $data);
        } else {
            return $this->_findNode($node->right, $data);
        }
    }

    public function getRoot() {
        return $this->root;
    }

    public function preorder($node) {
        if (empty($node)) {
            return;
        }
        echo $node->data . " ";
        $this->preorder($node->left);
        $this->preorder($node->right);
    }

    public function inorder($node) {
        if (empty($node)) {
            return;
        }
        $this->inorder($node->left);
        echo $node->data . " ";
        $this->inorder($node->right);
    }

    public function postorder($node) {
        if (empty($node)) {
            return;
        }
        $this->postorder($node->left);
        $this->postorder($node->right);
        echo $node->data . " ";
    }

    public function levelorder($node) {
        if (empty($node)) {
            return;
        }
        $queue = [$node];
        while (!empty($queue)) {
            $curr = array_shift($queue);
            echo $curr->data . " ";

            if (!empty($curr->left)) {
                $queue[] = $curr->left;
            }
            if (!empty($curr->right)) {
                $queue[] = $curr->right;
            }
        }
    }
}

$bst = new BinarySearchTree(40);
$root = $bst->getRoot();
$bst->insert(13);
$bst->insert(57);
$bst->insert(7);
$bst->insert(37);
$bst->insert(49);
$bst->insert(67);
$bst->insert(34);
$bst->insert(39);
$bst->insert(63);
$bst->insert(28);
$bst->insert(38);
$bst->insert(60);
$bst->insert(65);
$bst->insert(30);
$bst->insert(29);
$bst->insert(32);

$bst->preorder($root);
echo PHP_EOL;
$bst->inorder($root);
echo PHP_EOL;
$bst->postorder($root);
echo PHP_EOL;
$bst->levelorder($root);
echo PHP_EOL;
$bst->find(29);
$bst->find(32);

// $bst->delete(32);
// $bst->preorder($root);
// echo PHP_EOL;
// $bst->inorder($root);
// echo PHP_EOL;
// $bst->postorder($root);
// echo PHP_EOL;